SPRING-1 - initial commit

mblawat opened the pull request:
Initial commit that contais the whole excercise.

mblawat committed:
SPRING-1 - initial commit

mblawat committed:
Merge branch 'main' into feature/SPRING-1-initial-commit

dabr-grapeup commented on pom.xml:
```diff
@@ -0,0 +1,53 @@
+		<relativePath/> <!-- lookup parent from repository -->
+	</parent>
+	<groupId>com.grapeup.webapp</groupId>
+	<artifactId>testWebApp</artifactId>
```
Artfiacts are named using kebab-case, notice other dependencies.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/HomeController.java:
```diff
@@ -0,0 +1,23 @@
+    private String appVersion;
+
+    @GetMapping
+    public Map getStatus() {
```
You should include type of keys and values, otherwise if you would ever use this method it would require from you to cast it before you can operate on it.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/HomeController.java:
```diff
@@ -0,0 +1,23 @@
+
+    @GetMapping
+    public Map getStatus() {
+        Map map = new HashMap<String, String>();
```
There is factory method for Map with single element: `Collections.singletonMap()`, but in this Java version you also have available `Map.of()`, which I strongly recommend.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SessionsController.java:
```diff
@@ -0,0 +1,47 @@
+@RestController
+@RequestMapping("/api/v1/sessions")
+public class SessionsController {
+    @Autowired
```
Injection by constructor is recommended.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SessionsController.java:
```diff
@@ -0,0 +1,47 @@
+@RequestMapping("/api/v1/sessions")
+public class SessionsController {
+    @Autowired
+    private SessionRepository sessionRepository;
```
Should be `final` as well.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SessionsController.java:
```diff
@@ -0,0 +1,47 @@
+    public Session update(@PathVariable final Long id, @RequestBody Session session) {
+        // TODO: Validate body
+        Session existingSession = sessionRepository.getById(id);
+        BeanUtils.copyProperties(session, existingSession, "session_id");
```
Usually you would like to do that kind of stuff in service layer, but I get it for that simple example it's not needed. :)

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SpeakersController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @GetMapping
+    @RequestMapping("{id}")
```
You can pass `{id}` to `@GetMapping`.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SpeakersController.java:
```diff
@@ -0,0 +1,48 @@
+    @PutMapping("{id}")
+    public Speaker update(@PathVariable final Long id, @RequestBody Speaker speaker) {
+        // TODO: Validate body
+        Speaker existing_speaker = speakerRepository.getById(id);
```
Please use `camelCase`.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/models/Session.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.List;
+
+@Entity(name = "sessions")
+@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
```
You should never expose entity directly via API. Please consider creating DTO object for it.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/models/Session.java:
```diff
@@ -0,0 +1,68 @@
+public class Session {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long session_id;
```
`camelCase` please!

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/models/Session.java:
```diff
@@ -0,0 +1,68 @@
+            inverseJoinColumns = @JoinColumn(name = "speaker_id"))
+    private List<Speaker> speakers;
+
+    public Session() {}
```
You can check `lombok` library to generate constructor and getters/setters by just using annotations.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/repositories/SessionRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.grapeup.webapp.testWebApp.models.Session;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface SessionRepository extends JpaRepository<Session, Long> {
```
I suppose you don't have a reason to use `JpaRepository` over `CrudRepository` https://www.tutorialspoint.com/difference-between-crudrepository-and-jparepository-in-java

dabr-grapeup commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,10 @@
+spring.datasource.url=${DB_URL}
```
FYI Spring will automatically use environment variables if they are named the same way as properties, eg. SPRING_DATASOURCE_URL

dabr-grapeup commented on src/test/java/com/grapeup/webapp/testWebApp/TestWebAppApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class TestWebAppApplicationTests {
```
Where are your tests?! 😄

mblawat committed:
SPRING-1 - Exception handlers, CI improvements, Controller unit tests
1. Changed JPA repositories to CRUD ones.
2. Changed entity properties to be camelCase while retaining snake case for columns.
3. Added Exception handlers along with Session and Speaker exception for controllers.
4. Added unit tests for Sessions and Speaker controllers

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/controllers/SessionsController.java:
```diff
@@ -1,34 +1,43 @@
-        return sessionRepository.getById(id);
+        Optional<Session> existingSession = sessionRepository.findById(id);
+        if(existingSession.isPresent()) {
+            return existingSession.get();
```
There is nice method to simplify it by `ResponseEntity.of(<optional goes here>)`, controller advice is a great idea, but a bit overkill for `NOT_FOUND`.

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/exceptions/ControllerAdvisor.java:
```diff
@@ -0,0 +1,28 @@
+package com.grapeup.webapp.testWebApp.exceptions;
```
You can update package as well: `testWebApp`

dabr-grapeup commented on src/main/java/com/grapeup/webapp/testWebApp/exceptions/SessionException.java:
```diff
@@ -0,0 +1,9 @@
+package com.grapeup.webapp.testWebApp.exceptions;
+
+public class SessionException extends RuntimeException {
```
Another thing which you could try for such simple use case is just annotate exception with `@ResponseStatus`, Spring will serialize exception to response with given HTTP status.

dabr-grapeup reviewed (commented):
Hmm I'm not too familiar with `JSONAssert` and `JacksonTester`, but it looks ok. It would be good to add integration test with database in memory, on top of that I suggest to use `TestRestTemplate` instead of `MockMvc`. To be honest I'm not the fun of second one and not every project will use it. 😄
